### PR TITLE
Add wallet database

### PR DESCRIPTION
**Pull request by raphjaph**

### CODE REVIEW

File: src/subcommand/wallet.rs [hunk 1]
```diff
@@ -51,6 +54,8 @@ pub(crate) enum Subcommand {
   Mint(mint::Mint),
   #[command(about = "Generate receive address")]
   Receive(receive::Receive),
+  #[command(about = "Resume any pending etchings")]
+  Resume,
```
Comment (rot13maxi) [inline src/subcommand/wallet.rs:58]: 🎉

File: src/wallet.rs [hunk 2]
```diff
@@ -509,6 +544,68 @@ impl Wallet {
     self.settings.integration_test()
   }
 
+  pub(crate) fn wait_for_maturation(
+    &self,
+    rune: &Rune,
+    commit_tx: Transaction,
```
Comment (casey) [inline src/wallet.rs:550]: Maybe just `commit` and `reveal`?

File: src/wallet.rs [hunk 3]
```diff
@@ -509,6 +544,68 @@ impl Wallet {
...
+    eprintln!(
+      "Waiting for rune commitment {} to mature…",
+      commit_tx.txid()
+    );
+
+    self.store(rune, &commit_tx, &reveal_tx, output.clone())?;
+
+    loop {
+      if SHUTTING_DOWN.load(atomic::Ordering::Relaxed) {
+        return Err(anyhow!("cancelled"));
```
Comment (casey) [inline src/wallet.rs:563]: I think this shouldn't be an error.

File: src/wallet.rs [hunk 4]
```diff
@@ -509,6 +544,68 @@ impl Wallet {
...
+    commit_tx: Transaction,
+    reveal_tx: Transaction,
+    output: batch::Output,
+  ) -> Result<batch::Output> {
+    eprintln!(
+      "Waiting for rune commitment {} to mature…",
+      commit_tx.txid()
+    );
+
+    self.store(rune, &commit_tx, &reveal_tx, output.clone())?;
```
Comment (casey) [inline src/wallet.rs:559]: Maybe like `save_etching`?

File: src/wallet.rs [hunk 5]
```diff
@@ -509,6 +544,68 @@ impl Wallet {
...
+      Ok(txid) => txid,
+      Err(err) => {
+        return Err(anyhow!(
+          "Failed to send reveal transaction: {err}\nCommit tx {} will be recovered once mined",
+          commit_tx.txid()
+        ))
+      }
+    };
+
+    self.clear(rune)?;
```
Comment (casey) [inline src/wallet.rs:604]: Maybe `clear_etching`?

File: src/wallet.rs [hunk 6]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+    let progress_bar = Mutex::new(None);
+    let integration_test = settings.integration_test();
+
+    let durability = if cfg!(test) {
+      redb::Durability::None
+    } else {
+      redb::Durability::Immediate
+    };
+
+    let repair_callback = move |progress: &mut RepairSession| {
```
Comment (casey) [inline src/wallet.rs:787]: Let's deduplicate this

File: src/wallet.rs [hunk 7]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+
+        let mut tx = database.begin_write()?;
+
+        tx.set_durability(durability);
+
+        tx.open_table(RUNE_TO_INFO)?;
+
+        {
+          let mut statistics = tx.open_table(STATISTICS)?;
+          statistics.insert(&Statistic::Schema.key(), &SCHEMA_VERSION)?;
```
Comment (casey) [inline src/wallet.rs:855]:
```suggestion
          tx.open_table(STATISTICS)?.insert(&Statistic::Schema.key(), &SCHEMA_VERSION)?;
```

File: src/wallet.rs [hunk 8]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+        }
+
+        tx.commit()?;
+
+        database
+      }
+      Err(error) => bail!("failed to open index: {error}"),
```
Comment (casey) [inline src/wallet.rs:862]:
```suggestion
      Err(error) => bail!("failed to open wallet database: {error}"),
```

File: src/wallet.rs [hunk 9]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+    };
+
+    Ok(database)
+  }
+
+  fn begin_read(&self) -> Result<ReadTransaction> {
```
Comment (casey) [inline src/wallet.rs:868]: Let's remove this one

File: src/wallet.rs [hunk 10]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+    Ok(self.database.begin_read()?)
+  }
+
+  fn begin_write(&self) -> Result<WriteTransaction> {
```
Comment (casey) [inline src/wallet.rs:872]: Let's nuke this and not bother setting the durability.

File: src/wallet.rs [hunk 11]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+      }
+      .store(),
+    )?;
+
+    wtx.commit()?;
+
+    Ok(())
+  }
+
+  pub(crate) fn retrieve(&self, rune: Rune) -> Result<Option<ResumeEntry>> {
```
Comment (casey) [inline src/wallet.rs:907]:
```suggestion
  pub(crate) fn load_etching(&self, rune: Rune) -> Result<Option<ResumeEntry>> {
```

File: src/wallet.rs [hunk 12]
```diff
@@ -8,19 +8,51 @@ use {
...
   transaction_builder::TransactionBuilder,
 };
 
 pub mod batch;
+pub mod entry;
 pub mod transaction_builder;
 
+const SCHEMA_VERSION: u64 = 1;
+
+define_table! { RUNE_TO_INFO, u128, ResumeEntryValue }
```
Comment (casey) [inline src/wallet.rs:36]: Maybe `RUNE_TO_ETCHING`?

File: src/wallet.rs [hunk 13]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+  pub(crate) fn clear(&self, rune: &Rune) -> Result {
+    let wtx = self.begin_write()?;
+
+    wtx.open_table(RUNE_TO_INFO)?.remove(rune.0)?;
+    wtx.commit()?;
+
+    Ok(())
+  }
+
+  pub(crate) fn pending(&self) -> Result<Vec<(Rune, ResumeEntry)>> {
```
Comment (casey) [inline src/wallet.rs:927]:
```suggestion
  pub(crate) fn pending_etchings(&self) -> Result<Vec<(Rune, ResumeEntry)>> {
```

File: src/wallet.rs [hunk 14]
```diff
@@ -665,4 +762,179 @@ impl Wallet {
...
+
+    Ok(
+      rtx
+        .open_table(RUNE_TO_INFO)?
+        .get(rune.0)?
+        .map(|result| ResumeEntry::load(result.value())),
+    )
+  }
+
+  pub(crate) fn clear(&self, rune: &Rune) -> Result {
```
Comment (casey) [inline src/wallet.rs:918]:
```suggestion
  pub(crate) fn clear_etching(&self, rune: Rune) -> Result {
```

File: src/wallet/entry.rs [hunk 15]
```diff
@@ -0,0 +1,104 @@
+use super::*;
+
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct ResumeEntry {
```
Comment (casey) [inline src/wallet/entry.rs:4]:
```suggestion
pub struct EtchingEntry {
```

File: src/wallet/entry.rs [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+);
+
+impl Entry for ResumeEntry {
+  type Value = ResumeEntryValue;
+
+  fn load((commit, reveal, output): ResumeEntryValue) -> Self {
+    Self {
+      commit: consensus::encode::deserialize::<Transaction>(&commit).unwrap(),
+      reveal: consensus::encode::deserialize::<Transaction>(&reveal).unwrap(),
+      output: serde_json::from_str(std::str::from_utf8(&output).unwrap()).unwrap(),
```
Comment (casey) [inline src/wallet/entry.rs:23]:
```suggestion
      output: serde_json::from_slice(&output)).unwrap(),
```

File: tests/lib.rs [hunk 17]
```diff
@@ -196,7 +197,7 @@ fn batch(core: &mockcore::Handle, ord: &TestServer, batchfile: batch::File) -> E
     .read_line(&mut buffer)
     .unwrap();
 
-  assert_eq!(buffer, "Waiting for rune commitment to mature…\n");
+  assert_regex_match!(buffer, "Waiting for rune commitment .* to mature…\n");
```
Comment (casey) [inline tests/lib.rs:200]: Maybe check for the txid, like with `[[:xdigit]]{???}`

**Review by casey — CHANGES_REQUESTED** [review]
SO CLOSE

**Review by casey — APPROVED** [review]
LGTM

**Comment by Sidey123** [comment]
Dode the runes wallet down?

**Comment by Frank-Allen** [comment]
What kind of wallet do you use for Rune?

**Pull request merged**